godtossing/Pos/Ssc/GodTossing/Core/Types.hs

```diff
@@ -37,42 +38,50 @@ module Pos.Ssc.GodTossing.Core.Types
        , NodeSet
        ) where
 
+import           Control.Lens        (each, traverseOf)
 import           Data.Hashable       (Hashable (..))
+import           Data.HashMap.Strict (HashMap)
 import qualified Data.HashMap.Strict as HM
 import qualified Data.Text.Buildable
 import           Formatting          (bprint, build, int, (%))
 import           Universum
 
-import           Pos.Binary.Class    (AsBinary (..))
+import           Pos.Binary.Class    (AsBinary (..), fromBinaryM, serialize')
 import           Pos.Binary.Core     ()
 import           Pos.Core.Address    (addressHash)
 import           Pos.Core.Types      (EpochIndex, StakeholderId)
-import           Pos.Crypto          (EncShare, Hash, PublicKey, Secret, SecretKey,
-                                      SecretProof, SecretSharingExtra, Share,
-                                      SignTag (SignVssCert), Signature, VssPublicKey,
-                                      checkSig, sign, toPublic)
+import           Pos.Crypto          (DecShare, EncShare, Hash, PublicKey, Secret,
+                                      SecretKey, SecretProof, SignTag (SignVssCert),
+                                      Signature, VssPublicKey, checkSig, sign, toPublic)
 
 type NodeSet = HashSet StakeholderId
 
 ----------------------------------------------------------------------------
 -- Commitments
 ----------------------------------------------------------------------------
 
--- | Commitment is a message generated during the first stage of
--- GodTossing. It contains encrypted shares and proof of secret.
--- Invariant which must be ensured: commShares is not empty.
+-- | Commitment is a message generated during the first stage of GodTossing.
+-- It contains encrypted shares and proof of secret.
+--
+-- There can be more than one share generated for a single participant.
 data Commitment = Commitment
-    { commExtra  :: !(AsBinary SecretSharingExtra)
-    , commProof  :: !(AsBinary SecretProof)
-    , commShares :: !(HashMap (AsBinary VssPublicKey) (NonEmpty (AsBinary EncShare)))
+    { commProof  :: !SecretProof
+    , commShares :: !(HashMap (AsBinary VssPublicKey)
+                              (NonEmpty (AsBinary EncShare)))
     } deriving (Show, Eq, Generic)
 
 instance NFData Commitment
 instance Hashable Commitment
 
+-- | Get commitment shares.
+getCommShares :: Commitment -> Maybe [(VssPublicKey, NonEmpty EncShare)]
+getCommShares =
+    traverseOf (each . _1) fromBinaryM <=<          -- decode keys
+    traverseOf (each . _2 . each) fromBinaryM .     -- decode shares
+    HM.toList . commShares
+
 instance Ord Commitment where
-    compare = comparing commExtra <>
-              comparing commProof <>
+    compare = comparing (serialize' . commProof) <>
```

Comment:
I haven't checked whether the serialization is fast enough.

We need this instance because during CBOR decoding we need `Ord` to enforce canonicity, unfortunately. The best solution is to enforce canonicity on CBOR level but with `cborg` it's hard to do.